File: bindings/nim/src/proven/safe_network.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe network operations for IP address validation and classification.
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

type
  IPv4* = object
    ## Represents an IPv4 address.
    a*, b*, c*, d*: uint8

proc `$`*(ip: IPv4): string =
  ## Convert IPv4 to string representation.
  $ip.a & "." & $ip.b & "." & $ip.c & "." & $ip.d

proc parseIPv4*(address: string): Option[IPv4] =
  ## Parse an IPv4 address string.
  if address.len == 0:
    return none(IPv4)
  let res = provenNetworkParseIpv4(unsafeAddr address[0], csize_t(address.len))
  if res.status == PROVEN_OK:
    return some(IPv4(
      a: res.address.octets[0],
      b: res.address.octets[1],
      c: res.address.octets[2],
      d: res.address.octets[3]
    ))
  none(IPv4)

proc isValidIPv4*(address: string): bool =
  ## Check if a string is a valid IPv4 address.
  parseIPv4(address).isSome

proc isPrivate*(address: string): bool =
  ## Check if an IPv4 address is in a private range.
  if address.len == 0:
    return false
  let res = provenNetworkParseIpv4(unsafeAddr address[0], csize_t(address.len))
  if res.status != PROVEN_OK:
    return false
  provenNetworkIpv4IsPrivate(res.address)

proc isLoopback*(address: string): bool =
  ## Check if an IPv4 address is a loopback address (127.0.0.0/8).
  if address.len == 0:
    return false
  let res = provenNetworkParseIpv4(unsafeAddr address[0], csize_t(address.len))
  if res.status != PROVEN_OK:
    return false
  provenNetworkIpv4IsLoopback(res.address)

proc isPublic*(address: string): bool =
  ## Check if an IPv4 address is public (not private or loopback).
  isValidIPv4(address) and not isPrivate(address) and not isLoopback(address)

proc formatIPv4*(ip: IPv4): string =
  ## Format an IPv4 address as a string.
  $ip

File: bindings/lua/proven/safe_network.lua
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Safe network operations for IP address validation and classification.
--- @module proven.safe_network

local safe_network = {}

--- Parse an IPv4 address string.
--- @param address string
--- @return table|nil Returns {a, b, c, d} or nil if invalid
function safe_network.parse_ipv4(address)
    local parts = {}
    for part in address:gmatch("([^%.]+)") do
        local num = tonumber(part)
        if not num or num < 0 or num > 255 or math.floor(num) ~= num then
            return nil
        end
        table.insert(parts, num)
    end

    if #parts ~= 4 then
        return nil
    end

    return {
        a = parts[1],
        b = parts[2],
        c = parts[3],
        d = parts[4]
    }
end

--- Check if a string is a valid IPv4 address.
--- @param address string
--- @return boolean
function safe_network.is_valid_ipv4(address)
    return safe_network.parse_ipv4(address) ~= nil
end

--- Check if an IPv4 address is in a private range.
--- @param address string
--- @return boolean
function safe_network.is_private(address)
    local ip = safe_network.parse_ipv4(address)
    if not ip then
        return false
    end

    -- 10.0.0.0/8
    if ip.a == 10 then
        return true
    end

    -- 172.16.0.0/12
    if ip.a == 172 and ip.b >= 16 and ip.b <= 31 then
        return true
    end

    -- 192.168.0.0/16
    if ip.a == 192 and ip.b == 168 then
        return true
    end

    return false
end

--- Check if an IPv4 address is a loopback address (127.0.0.0/8).
--- @param address string
--- @return boolean
function safe_network.is_loopback(address)
    local ip = safe_network.parse_ipv4(address)
    if not ip then
        return false
    end
    return ip.a == 127
end

--- Check if an IPv4 address is public (not private or loopback).
--- @param address string
--- @return boolean
function safe_network.is_public(address)
    return safe_network.is_valid_ipv4(address)
        and not safe_network.is_private(address)
        and not safe_network.is_loopback(address)
end

--- Format an IPv4 address as a string.
--- @param ip table {a, b, c, d}
--- @return string
function safe_network.format_ipv4(ip)
    return string.format("%d.%d.%d.%d", ip.a, ip.b, ip.c, ip.d)
end

return safe_network
